File: 282_expression_add_operators.rb
=begin
Given a string that contains only digits 0-9 and a target value, return all possibilities to add binary operators (not unary) +, -, or * between the digits so they evaluate to the target value.

Example 1:

Input: num = "123", target = 6
Output: ["1+2+3", "1*2*3"] 
Example 2:

Input: num = "232", target = 8
Output: ["2*3+2", "2+3*2"]
Example 3:

Input: num = "105", target = 5
Output: ["1*0+5","10-5"]
Example 4:

Input: num = "00", target = 0
Output: ["0+0", "0-0", "0*0"]
Example 5:

Input: num = "3456237490", target = 9191
Output: []
=end

# @param {String} num
# @param {Integer} target
# @return {String[]}
def add_operators(num, target)
    answers = []
    recurse(0, 0, 0, 0, "", num, target, answers)
    
    return answers
end

def recurse(index, prev_operand, current_operand, value, string, num, target, answers)
    if index == num.length
        if value == target && current_operand == 0
            answers << string[1..-1]
        end
        return
    end
    
    current_operand = current_operand*10 + num[index].to_i
    str_op = current_operand.to_s
    
    if current_operand > 0
        recurse(index + 1, prev_operand, current_operand, value, string, num, target, answers)
    end
    string.concat('+').concat(str_op)
    recurse(index + 1, current_operand, 0, value + current_operand, string, num, target, answers)
    string.chomp!(str_op).chomp!('+')
    if !string.empty?
        string.concat('-').concat(str_op)
        recurse(index + 1, -current_operand, 0, value - current_operand, string, num, target, answers)
        string.chomp!(str_op).chomp!('-')
        
        string.concat('*').concat(str_op)
        recurse(index + 1, current_operand * prev_operand, 0, value - prev_operand + (current_operand * prev_operand), string, num, target, answers)
        string.chomp!(str_op).chomp!('*')
    end
end

File: 282_expression_add_operators.java
/*
Given a string num that contains only digits and an integer target, return all possibilities to add the binary operators '+', '-', or '*' between the digits of num so that the resultant expression evaluates to the target value.

 

Example 1:

Input: num = "123", target = 6
Output: ["1*2*3","1+2+3"]
Example 2:

Input: num = "232", target = 8
Output: ["2*3+2","2+3*2"]
Example 3:

Input: num = "105", target = 5
Output: ["1*0+5","10-5"]
Example 4:

Input: num = "00", target = 0
Output: ["0*0","0+0","0-0"]
Example 5:

Input: num = "3456237490", target = 9191
Output: []
 

Constraints:

1 <= num.length <= 10
num consists of only digits.
-231 <= target <= 231 - 1
*/

class Solution {
    public ArrayList<String> answer;
    public String digits;
    public long target;
    
    public void recurse(
        int index, long previousOperand, long currentOperand, long value, ArrayList<String> ops) {
        String nums = this.digits;
        if (index == nums.length()) {
            if (value == this.target && currentOperand == 0) {
                StringBuilder sb = new StringBuilder();
                ops.subList(1, ops.size()).forEach(v -> sb.append(v));
                this.answer.add(sb.toString());
            }
            return;
        }
        
        currentOperand = currentOperand * 10 + Character.getNumericValue(nums.charAt(index));
        String current_val_rep = Long.toString(currentOperand);
        int length = nums.length();
        
        if (currentOperand > 0) {
            recurse(index + 1, previousOperand, currentOperand, value, ops);
        }
        
        ops.add("+");
        ops.add(current_val_rep);
        recurse(index + 1, currentOperand, 0, value + currentOperand, ops);
        ops.remove(ops.size() - 1);
        ops.remove(ops.size() - 1);
        
        if (ops.size() > 0) {
            ops.add("-");
            ops.add(current_val_rep);
            recurse(index + 1, -currentOperand, 0, value - currentOperand, ops);
            ops.remove(ops.size() - 1);
            ops.remove(ops.size() - 1);
            
            ops.add("*");
            ops.add(current_val_rep);
            recurse(
                index + 1,
                currentOperand * previousOperand,
                0,
                value - previousOperand + (currentOperand * previousOperand),
                ops);
            ops.remove(ops.size() - 1);
            ops.remove(ops.size() - 1);
        }
    }
    
    public List<String> addOperators(String num, int target) {
        if (num.length() == 0) {
            return new ArrayList<String>();
        }
        
        this.target = target;
        this.digits = num;
        this.answer = new ArrayList<String>();
        this.recurse(0, 0, 0, 0, new ArrayList<String>());
        
        return this.answer;
    }
}

